src/Data/Array/NonEmpty.purs

```diff
@@ -0,0 +1,502 @@
+--
+-- Note that this is unsafe: if the transform returns an empty array, this can
+-- explode at runtime.
+unsafeAdapt :: forall a b. (Array a -> Array b) -> NonEmptyArray a -> NonEmptyArray b
+unsafeAdapt f = NonEmptyArray <<< adaptAny f
+
+-- | Internal - adapt an Array transform with argument to NonEmptyArray
+--
+-- Note that this is unsafe: if the transform returns an empty array, this can
+-- explode at runtime.
+unsafeAdapt' :: forall a b c. (a -> Array b -> Array c) -> a -> NonEmptyArray b -> NonEmptyArray c
+unsafeAdapt' f = unsafeAdapt <<< f
+
+-- | Internal - adapt an Array transform with two arguments to NonEmptyArray
+--
+-- Note that this is unsafe: if the transform returns an empty array, this can
+-- explode at runtime.
+unsafeAdapt'' :: forall a b c d. (a -> b -> Array c -> Array d) -> a -> b -> NonEmptyArray c -> NonEmptyArray d
+unsafeAdapt'' f = unsafeAdapt' <<< f
+
+-- | Internal - adapt an Array transform to NonEmptyArray,
+--   with polymorphic result.
+--
+-- Note that this is unsafe: if the transform returns an empty array, this can
+-- explode at runtime.
+adaptAny :: forall a b. (Array a -> b) -> NonEmptyArray a -> b
+adaptAny f = f <<< toArray
+
+-- | Internal - adapt an Array transform with argument to NonEmptyArray,
+--   with polymorphic result.
+adaptAny' :: forall a b c. (a -> Array b -> c) -> a -> NonEmptyArray b -> c
+adaptAny' f = adaptAny <<< f
+
+-- | Internal - adapt an Array transform with two arguments to NonEmptyArray,
+--   with polymorphic result.
+adaptAny'' :: forall a b c d. (a -> b -> Array c -> d) -> a -> b -> NonEmptyArray c -> d
+adaptAny'' f = adaptAny' <<< f
+
+-- | Internal - adapt Array functions returning Maybes to NonEmptyArray
+adaptMaybe :: forall a b. (Array a -> Maybe b) -> NonEmptyArray a -> b
+adaptMaybe f = unsafePartial $ fromJust <<< f <<< toArray
+
+fromArray :: forall a. Array a -> Maybe (NonEmptyArray a)
+fromArray xs
+  | A.length xs > 0 = Just (NonEmptyArray xs)
+  | otherwise       = Nothing
+
+-- | INTERNAL
+unsafeFromArray :: forall a. Array a -> NonEmptyArray a
+unsafeFromArray = NonEmptyArray
+
+fromNonEmpty :: forall a. NonEmpty Array a -> NonEmptyArray a
+fromNonEmpty (x :| xs) = cons' x xs
+
+toArray :: forall a. NonEmptyArray a -> Array a
+toArray (NonEmptyArray xs) = xs
+
+toNonEmpty :: forall a. NonEmptyArray a -> NonEmpty Array a
+toNonEmpty = uncons >>> \{head: x, tail: xs} -> x :| xs
+
+fromFoldable :: forall f a. Foldable f => f a -> Maybe (NonEmptyArray a)
+fromFoldable = fromArray <<< A.fromFoldable
+
+fromFoldable1 :: forall f a. Foldable1 f => f a -> NonEmptyArray a
+fromFoldable1 = unsafeFromArray <<< A.fromFoldable
+
+toUnfoldable :: forall f a. Unfoldable f => NonEmptyArray a -> f a
+toUnfoldable = adaptAny A.toUnfoldable
+
+singleton :: forall a. a -> NonEmptyArray a
+singleton = NonEmptyArray <<< A.singleton
+
+range :: Int -> Int -> Maybe (NonEmptyArray Int)
+range x y = fromArray $ A.range x y
```

Comment:
Maybe like with `replicate` this should always produce at least 1 value? Actually, isn't that basically guaranteed anyway? Unless "reverse-ranges" don't work, but I don't see why they couldn't (although obviously that'd be more of a change as we'd want it to behave consistently with normal arrays).